lib/ansible/modules/windows/win_regedit.ps1

```diff
@@ -21,313 +22,326 @@
+    try {
+        $value = (Get-Item -Path $path).GetValue($name)
+        if ($value -eq $null) {
+            return $false
+        } else {
+            return $true
+        }
+    } catch {
+        # key didn't exist so the property mustn't
+        return $false
+    }
 }
 
-# Special case handling for the path's default property.
-if ($name -eq "" -or ($name -ne $null -and $name.ToLower() -eq "(default)")) {
-    # Apparently, "(default)" cannot be of type expandstring, do it ourselves
-    if ($type -eq "expandstring" -and $data -ne $null -and $data -is [string]) {
-        $data = Expand-Environment($data)
+Function Compare-RegistryProperties($existing, $new) {
+    $mismatch = $false
+    if ($existing -is [Array]) {
+        if ((Compare-Object -ReferenceObject $existing -DifferenceObject $new -SyncWindow 0).Length -ne 0) {
+            $mismatch = $true
+        }
+    } else {
+        if ($existing -cne $new) {
+            $mismatch = $true
+        }
     }
-    $name = "(default)"
-    $type = "string"
+
+    $mismatch
 }
 
-# Support REG_NONE with empty value
-# FIXME: REG_NONE support is not idempotent yet
-if ($type -eq "none" -or $data -eq $null) {
-    $data = New-Object byte[] 0
-#    $data = ([byte[]] @())
+# convert property names "" to $null as "" refers to (Default)
+if ($name -eq "") {
+    $name = $null
 }
 
-if ($state -eq "present") {
+# convert the data to the required format
+if ($type -in @("binary", "none")) {
+    if ($data -eq $null) {
+        $data = ""
+    }
 
-    if ((Test-Path $path) -and $name -ne $null) {
+    # convert the data from string to byte array if in hex: format
+    if ($data -is [String]) {
+        $data = [byte[]](Convert-RegExportHexStringToByteArray -string $data)
+    } elseif ($data -is [Int]) {
+        if ($data -gt 255) {
+            Fail-Json $result "cannot convert binary data '$data' to byte array, please specify this value as a yaml byte array or a comma separate hex value string"
+        }
+        $data = [byte[]]@([byte]$data)
+    } elseif ($data -is [Array]) {
+        $data = [byte[]]$data
+    }
+} elseif ($type -in @("dword", "qword")) {
+    # dwords and qwords don't have the concept of a null value, will set to 0
+    if ($data -eq $null) {
+        $data = 0
+    }
 
-        if (Test-ValueData -Path $path -Name $name) {
+    if ($data -is [String]) {
+        # if the data is a string we need to convert it to an unsigned int64
+        # it needs to be unsigned as Ansible passes in an unsigned value while
+        # powershell uses a signed data type. The value will then be converted
+        # below
+        $data = [UInt64]$data
+    }
 
-            # Handle binary data
-            $old_data = (Get-ItemProperty -Path $path | Select-Object -ExpandProperty $name)
+    # convert the input variable from Int64 to Int32 when dealing with dwords
+    if ($type -eq "dword" -and $data -gt [Int32]::MaxValue) {
```

Comment:
Ah OK, now I see... Yeah, looks like that's the best way short of dropping into C#, since Powershell doesn't appear to expose a way to enable `unchecked` narrowing conversions.